borsh/decode.go

```
@@ -0,0 +1,127 @@
+	return BorshDecoder{
+		data:   data,
+		offset: 0,
+	}
+}
+
+// Finished returns a bool on whether the decoding has finished
+func (decoder *BorshDecoder) Finished() bool {
+	return decoder.offset == uint32(len(decoder.data))
+}
+
+// DecodeU8 deodes the input bytes and returns the corresponding
+// `uint8` value and any errors
+func (decoder *BorshDecoder) DecodeU8() (uint8, error) {
+	if uint32(len(decoder.data)) < decoder.offset+1 {
+		return 0, errors.New("Borsh: out of range")
+	}
+	val := uint8(decoder.data[decoder.offset])
+	decoder.offset++
+	return val, nil
+}
+
+// DecodeU16 deodes the input bytes and returns the corresponding
+// `uint8` value and any errors
+func (decoder *BorshDecoder) DecodeU16() (uint16, error) {
+	if uint32(len(decoder.data)) < decoder.offset+2 {
+		return 0, errors.New("Borsh: out of range")
+	}
+	val := binary.LittleEndian.Uint16(decoder.data[decoder.offset : decoder.offset+2])
+	decoder.offset += 2
+	return val, nil
+}
+
+// DecodeU32 deodes the input bytes and returns the corresponding
+// `uint32` value and any errors
+func (decoder *BorshDecoder) DecodeU32() (uint32, error) {
+	if uint32(len(decoder.data)) < decoder.offset+4 {
+		return 0, errors.New("Borsh: out of range")
+	}
+	val := binary.LittleEndian.Uint32(decoder.data[decoder.offset : decoder.offset+4])
+	decoder.offset += 4
+	return val, nil
+}
+
+// DecodeU64 deodes the input bytes and returns the corresponding
+// `uint64` value and any errors
+func (decoder *BorshDecoder) DecodeU64() (uint64, error) {
+	if uint32(len(decoder.data)) < decoder.offset+8 {
+		return 0, errors.New("Borsh: out of range")
+	}
+	val := binary.LittleEndian.Uint64(decoder.data[decoder.offset : decoder.offset+8])
+	decoder.offset += 8
+	return val, nil
+}
+
+// DecodeSigned8 deodes the input bytes and returns the corresponding signed
+// `int8` value and any errors
+func (decoder *BorshDecoder) DecodeSigned8() (int8, error) {
+	unsigned, _ := decoder.DecodeU8()
```

why do you ignore this error?